File: Data Structures & Algorithms/house-robber-ii/submission-3.java
class Solution {
    public int rob(int[] nums) {
        if(nums.length==1) return nums[0];
        int n = nums.length;
        return Math.max(helper(nums,0,n-2), helper(nums,1,n-1));
    }
    public int helper(int[] nums, int l, int r) {
        if (l == r) return nums[l];
        int x = nums[l];
        int y = Math.max(nums[l+1], nums[l]);
        for(int i=l+2; i<=r; i++) {
            int tmp = Math.max(nums[i]+x, y);
            x = y;
            y = tmp;
        }
        return y;
    }
}

File: Data Structures & Algorithms/maximum-product-subarray/submission-3.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        currmax = nums[0]
        currmin = nums[0]
        ans = nums[0]
        for i in range(1,len(nums)):
            tmp = currmax*nums[i]
            currmax = max(nums[i], currmax*nums[i], currmin*nums[i])
            currmin = min(nums[i], tmp, currmin*nums[i])
            ans = max(ans,currmax)
        return ans

# max = -3 (max till now / min till now * curr)
# min = -24 (min of num or max * num)
# ans = 8

# -24, -3, -24